File: packages/llmvault-mobile/android/src/main/java/io/llmvault/mobile/SSEParser.kt
package io.llmvault.mobile

import org.json.JSONObject

// ---------------------------------------------------------------------------
// Event types
// ---------------------------------------------------------------------------

/** Parsed SSE event types matching the SecureRelay event interface. */
sealed class SSEEvent {
    data class Delta(val text: String) : SSEEvent()
    data class Card(val card: Map<String, Any>) : SSEEvent()
    data class Done(val usage: StreamUsage?) : SSEEvent()
    data class Error(val message: String) : SSEEvent()
}

data class StreamUsage(
    val promptTokens: Int,
    val completionTokens: Int,
)

// ---------------------------------------------------------------------------
// Parser
// ---------------------------------------------------------------------------

/**
 * Parses Server-Sent Events from an LLM provider HTTP response body.
 * Mirrors the logic from llmvault-extension streamManager.ts and the Swift SSEParser.
 *
 * Not thread-safe — callers must synchronise access if feeding from multiple threads.
 */
class SSEParser(private val isAnthropic: Boolean) {

    private var buffer = ""

    /**
     * Feed a raw chunk of text from the HTTP stream.
     * Returns zero or more fully-parsed [SSEEvent]s.
     */
    fun feed(chunk: String): List<SSEEvent> {
        buffer += chunk

        // Split on newline without discarding the trailing empty element that
        // indicates a complete line — mirrors Swift split(omittingEmptySubsequences: false).
        val lines = buffer.split("\n")

        // If the buffer ends with a newline every line is complete; otherwise the
        // last element is an incomplete line that must stay in the buffer.
        if (buffer.endsWith("\n")) {
            buffer = ""
        } else {
            buffer = lines.last() ?: ""
        }

        val completedLines = if (buffer.isEmpty()) lines else lines.dropLast(1)

        val events = mutableListOf<SSEEvent>()
        for (line in completedLines) {
            val trimmed = line.trim()
            if (trimmed.isEmpty() || trimmed == "data: [DONE]") continue
            if (!trimmed.startsWith("data: ")) continue

            val jsonStr = trimmed.removePrefix("data: ")
            val json = runCatching { JSONObject(jsonStr) }.getOrNull() ?: continue

            if (isAnthropic) {
                events.addAll(parseAnthropicEvent(json))
            } else {
                events.addAll(parseOpenAiEvent(json))
            }
        }

        return events
    }

    /** Reset the parser state for reuse. */
    fun reset() {
        buffer = ""
    }

    // -----------------------------------------------------------------------
    // OpenAI SSE parsing
    // -----------------------------------------------------------------------

    /** Mirrors streamManager.ts parseOpenAiStreamEvent */
    private fun parseOpenAiEvent(data: JSONObject): List<SSEEvent> {
        val choices = data.optJSONArray("choices") ?: return emptyList()
        if (choices.length() == 0) return emptyList()
        val choice = choices.optJSONObject(0) ?: return emptyList()

        val events = mutableListOf<SSEEvent>()
        val delta = choice.optJSONObject("delta")

        // Text delta
        val content = delta?.optString("content", null)
        if (!content.isNullOrEmpty()) {
            events.add(SSEEvent.Delta(content))
        }

        // finish_reason present (may be null JSON value — optString returns "null" string for that)
        val finishReason = choice.opt("finish_reason")
        if (finishReason != null && finishReason != JSONObject.NULL) {
            val usageObj = data.optJSONObject("usage")
            val streamUsage = usageObj?.let {
                StreamUsage(
                    promptTokens = it.optInt("prompt_tokens", 0),
                    completionTokens = it.optInt("completion_tokens", 0),
                )
            }
            events.add(SSEEvent.Done(streamUsage))
        }

        return events
    }

    // -----------------------------------------------------------------------
    // Anthropic SSE parsing
    // -----------------------------------------------------------------------

    /** Mirrors streamManager.ts parseAnthropicStreamEvent */
    private fun parseAnthropicEvent(data: JSONObject): List<SSEEvent> {
        val eventType = data.optString("type", null) ?: return emptyList()
        val events = mutableListOf<SSEEvent>()

        when (eventType) {
            "content_block_delta" -> {
                val delta = data.optJSONObject("delta") ?: return emptyList()
                if (delta.optString("type") == "text_delta") {
                    val text = delta.optString("text", null)
                    if (!text.isNullOrEmpty()) {
                        events.add(SSEEvent.Delta(text))
                    }
                }
            }

            "message_delta" -> {
                val usageObj = data.optJSONObject("usage")
                val streamUsage = usageObj?.let {
                    StreamUsage(
                        // Anthropic uses input_tokens / output_tokens in message_delta
                        promptTokens = it.optInt("input_tokens", 0),
                        completionTokens = it.optInt("output_tokens", 0),
                    )
                }
                events.add(SSEEvent.Done(streamUsage))
            }

            "error" -> {
                val errorObj = data.optJSONObject("error")
                val message = errorObj?.optString("message", null)
                if (!message.isNullOrEmpty()) {
                    events.add(SSEEvent.Error(message))
                }
            }

            // Ignore: ping, message_start, content_block_start, content_block_stop, message_stop, etc.
            else -> Unit
        }

        return events
    }
}

File: packages/llmvault-mobile/ios/Sources/SecureRelayPlugin/SSEParser.swift
import Foundation

/// Parsed SSE event types matching the SecureRelay event interface.
enum SSEEvent {
    case delta(text: String)
    case card(card: [String: Any])
    case done(usage: StreamUsage?)
    case error(message: String)
}

struct StreamUsage {
    let promptTokens: Int
    let completionTokens: Int
}

/// Parses Server-Sent Events from LLM provider HTTP response body.
/// Mirrors the logic from llmvault-extension streamManager.ts.
final class SSEParser {

    private var buffer = ""
    private let isAnthropic: Bool

    init(isAnthropic: Bool) {
        self.isAnthropic = isAnthropic
    }

    /// Feed raw data from the HTTP stream and extract SSE events.
    func feed(_ chunk: String) -> [SSEEvent] {
        buffer += chunk
        let lines = buffer.split(separator: "\n", omittingEmptySubsequences: false)

        // Keep the last incomplete line in the buffer
        if buffer.hasSuffix("\n") {
            buffer = ""
        } else {
            buffer = String(lines.last ?? "")
        }

        var events: [SSEEvent] = []
        let completedLines = buffer.isEmpty ? lines : lines.dropLast()

        for line in completedLines {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            if trimmed.isEmpty || trimmed == "data: [DONE]" { continue }
            guard trimmed.hasPrefix("data: ") else { continue }

            let jsonStr = String(trimmed.dropFirst(6))
            guard let data = jsonStr.data(using: .utf8),
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                continue
            }

            if isAnthropic {
                events.append(contentsOf: parseAnthropicEvent(json))
            } else {
                events.append(contentsOf: parseOpenAiEvent(json))
            }
        }

        return events
    }

    // MARK: - OpenAI SSE Parsing

    /// Mirrors streamManager.ts parseOpenAiStreamEvent
    private func parseOpenAiEvent(_ data: [String: Any]) -> [SSEEvent] {
        guard let choices = data["choices"] as? [[String: Any]],
              let choice = choices.first else {
            return []
        }

        var events: [SSEEvent] = []
        let delta = choice["delta"] as? [String: Any]

        // Text delta
        if let content = delta?["content"] as? String {
            events.append(.delta(text: content))
        }

        // Finish reason → done
        if let _ = choice["finish_reason"] as? String {
            let usage = data["usage"] as? [String: Any]
            let streamUsage: StreamUsage?
            if let u = usage {
                streamUsage = StreamUsage(
                    promptTokens: u["prompt_tokens"] as? Int ?? 0,
                    completionTokens: u["completion_tokens"] as? Int ?? 0
                )
            } else {
                streamUsage = nil
            }
            events.append(.done(usage: streamUsage))
        }

        return events
    }

    // MARK: - Anthropic SSE Parsing

    /// Mirrors streamManager.ts parseAnthropicStreamEvent
    private func parseAnthropicEvent(_ data: [String: Any]) -> [SSEEvent] {
        guard let eventType = data["type"] as? String else { return [] }
        var events: [SSEEvent] = []

        switch eventType {
        case "content_block_delta":
            if let delta = data["delta"] as? [String: Any],
               let deltaType = delta["type"] as? String,
               deltaType == "text_delta",
               let text = delta["text"] as? String {
                events.append(.delta(text: text))
            }

        case "message_delta":
            let usage = data["usage"] as? [String: Any]
            let streamUsage: StreamUsage?
            if let u = usage {
                streamUsage = StreamUsage(
                    promptTokens: u["input_tokens"] as? Int ?? 0,
                    completionTokens: u["output_tokens"] as? Int ?? 0
                )
            } else {
                streamUsage = nil
            }
            events.append(.done(usage: streamUsage))

        case "error":
            if let errorData = data["error"] as? [String: Any],
               let message = errorData["message"] as? String {
                events.append(.error(message: message))
            }

        default:
            break // Ignore ping, message_start, content_block_start, etc.
        }

        return events
    }

    /// Reset the parser state for reuse.
    func reset() {
        buffer = ""
    }
}
